drivers/include/periph/hwcrypto.h

```diff
@@ -0,0 +1,394 @@
+} hwcrypto_cipher_t;
+
+/**
+ * @brief   Supported hashing algorithms.
+ *
+ * Define HAVE_HWCRYPTO_xxx in periph_cpu.h for each supported hashing
+ * algorithm.
+ *
+ * Extend at will, but also update drivers/periph_common/hwcrypto.c.
+ */
+typedef enum {
+    HWCRYPTO_MD4,                     /**< MD4 hashing algorithm */
+    HWCRYPTO_MD5,                     /**< MD5 hashing algorithm */
+    HWCRYPTO_SHA1,                    /**< SHA1 hashing algorithm */
+    HWCRYPTO_SHA3,                    /**< SHA3 hashing algorithm */
+    HWCRYPTO_SHA224,                  /**< SHA224 hashing algorithm */
+    HWCRYPTO_SHA256,                  /**< SHA256 hashing algorithm */
+    HWCRYPTO_SHA384,                  /**< SHA384 hashing algorithm */
+    HWCRYPTO_SHA512,                  /**< SHA512 hashing algorithm */
+    HWCRYPT_HASH_NUMOF                /**< number of hashes available */
+} hwcrypto_hash_t;
+
+/**
+ * @brief   Setup options for cipher and hash methods.
+ *
+ * Extend at will.
+ */
+typedef enum {
+    HWCRYPTO_OPT_KEY,
+    HWCRYPTO_OPT_IV,
+    HWCRYPTO_OPT_COUNTER,
+    HWCRYPTO_OPT_PADDING,
+} hwcrypto_opt_t;
+
+/**
+ * @brief   Supported cipher modes of operation, to be used with
+ *          @p HWCRYPTO_OPT_MODE.
+ *
+ * Extend at will.
+ */
+typedef enum {
+    HWCRYPTO_MODE_NONE,               /**< no mode */
+    HWCRYPTO_MODE_ECB,                /**< electronic codebook mode */
+    HWCRYPTO_MODE_CBC,                /**< cipher block chaining mode */
+    HWCRYPTO_MODE_CFB,                /**< cipher feedback mode */
+    HWCRYPTO_MODE_OFB,                /**< output feedback mode */
+    HWCRYPTO_MODE_CTR                 /**< counter mode */
+} hwcrypto_mode_t;
+
+/**
+ * @brief   Determine if the given hardware crypto peripheral supports the
+ *          given cipher.
+ *
+ * This method should not require the peripheral to be initialized. It will
+ * also not change its state.
+ *
+ * The method should return a deterministic result.
+ *
+ * @param[in] dev           the device
+ * @param[in] cipher        cipher to check
+ *
+ * @return                  nonzero value if supported
+ * @return                  0 if unsupported
+ */
+int hwcrypto_cipher_supported(hwcrypto_t dev, hwcrypto_cipher_t cipher);
+
+/**
+ * @brief   Determine if the given hardware crypto peripheral supports the
+ *          given hashing algorithm.
+ *
+ * This method should not require the peripheral to be initialized. It will
+ * also not change its state.
+ *
+ * The method should return a deterministic result.
+ *
+ * @param[in] dev           the device
+ * @param[in] hash          hashing algorithm to check
+ *
+ * @return                  nonzero value if supported
+ * @return                  0 if unsupported
+ */
+int hwcrypto_hash_supported(hwcrypto_t dev, hwcrypto_hash_t hash);
+
+/**
+ * @brief   Initialize the hardware crypto peripheral.
+ *
+ * The peripheral should be turned on in this method.
```

Comment:
Is it really required?

I have the impression this could be called from `periph_init` as spi or rtc can be initialized. So I would suggest the periph is not on after the call of this function, but setup to be ready for later usage. This is somehow the same discussion as for power on/off function, I think.

Reply:
Let's move it to `periph_init`